src/mr.rs

```diff
@@ -0,0 +1,499 @@
+use crate::{
+    device::{
+        ToCardCtrlRbDesc, ToCardCtrlRbDescCommon, ToCardCtrlRbDescUpdateMrTable,
+        ToCardCtrlRbDescUpdatePageTable,
+    },
+    responser::AcknowledgeBuffer,
+    types::{Key, MemAccessTypeFlag, PAGE_SIZE},
+    utils::{allocate_aligned_memory, deallocate_aligned_memory},
+    Device, Error, Pd,
+};
+use rand::RngCore as _;
+use std::{
+    hash::{Hash, Hasher},
+    mem, ptr,
+    sync::Arc,
+};
+
+const ACKNOWLEDGE_BUFFER_SLOT_CNT: usize = 1024;
+const ACKNOWLEDGE_BUFFER_SIZE: usize =
+    ACKNOWLEDGE_BUFFER_SLOT_CNT * AcknowledgeBuffer::ACKNOWLEDGE_BUFFER_SLOT_SIZE;
+
+/// Memory Region
+///
+/// User use `Device::alloc_mr(..)` to allocate a `Mr` and use `Device::dereg_mr(..)` to deallocate a `Mr`.
+#[derive(Debug, Clone, Copy)]
+pub struct Mr {
+    pub(crate) key: Key,
+}
+
+impl Mr {
+    /// Get the key of the Mr
+    #[must_use]
+    pub fn get_key(&self) -> Key {
+        self.key
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct MrCtx {
+    #[allow(unused)]
+    pub(crate) key: Key,
+    pub(crate) pd: Pd,
+    #[allow(unused)]
+    pub(crate) va: u64,
+    pub(crate) len: u32,
+    #[allow(unused)]
+    pub(crate) acc_flags: MemAccessTypeFlag,
+    pub(crate) pgt_offset: usize,
+    pub(crate) pg_size: u32,
+}
+
+#[derive(Debug)]
+pub(crate) struct MrPgt {
+    table: [u64; crate::MR_PGT_SIZE],
+    free_blk_list: *mut MrPgtFreeBlk,
+}
+
+struct MrPgtFreeBlk {
+    idx: usize,
+    len: usize,
+    prev: *mut Self,
+    next: *mut Self,
+}
+
+impl Device {
+    fn register_page_table(&self, addr: u64, length: u32, pg_size: u32) -> Result<usize, Error> {
+        let mut mr_pgt = self
+            .0
+            .mr_pgt
+            .lock()
```

Comment:
Again, the lock issue. I rethink about the lock verbose code issue, and find out that you'd better wrap it into a function for a data structure, returning the lock guard instead.

Reply:
I am not sure if I understand your meaning. Should I put expressions like `self.0,mr_pgt.lock()` into a method?